### PR TITLE
Create DM - Single Screen

### PR DESCRIPTION
**Pull request by onurays**
Currently, we have 3 different screens to invite users to new or existing rooms:
- Known Users Screen which contains users with who we have interaction
- Contact Book Screen
- Adding Users by Matrix Id Screen

We should reduce complexity and merge these 3 screens into 1 screen with sections:
- Recent
- Contacts
- Suggestions (search results)

merge_dm_screens

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/invite/InviteUsersToRoomActivity.kt [hunk 1]
```diff
@@ -101,26 +98,16 @@ class InviteUsersToRoomActivity : SimpleFragmentActivity() {
...
-            addFragmentToBackstack(R.id.container, ContactsBookFragment::class.java)
-        }
-    }
-
     override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
         super.onRequestPermissionsResult(requestCode, permissions, grantResults)
         if (allGranted(grantResults)) {
             if (requestCode == PERMISSION_REQUEST_CODE_READ_CONTACTS) {
-                doOnPostResume { addFragmentToBackstack(R.id.container, ContactsBookFragment::class.java) }
+                // TODO. Trigger VM
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/invite/InviteUsersToRoomActivity.kt:105]: TODO?

File: vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    @AssistedInject.Factory
+    interface Factory {
+        fun create(initialState: UserListViewState): UserListViewModel
+    }
+
+    companion object : MvRxViewModelFactory<UserListViewModel, UserListViewState> {
+        override fun create(viewModelContext: ViewModelContext, state: UserListViewState): UserListViewModel? {
+            return when (viewModelContext) {
+                is ActivityViewModelContext -> {
+                    when (viewModelContext.activity<FragmentActivity>()) {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt:76]: use when (val activity = viewModelContext.act...) in order to have smart cast?

File: vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+            is UserListAction.RemovePendingInvitee -> handleRemoveSelectedUser(action)
+            is UserListAction.OnlyBoundContacts    -> handleOnlyBoundContacts(action)
+        }.exhaustive
+    }
+
+    private fun handleSearchUsers(searchTerm: String) {
+        setState {
+            copy(searchTerm = searchTerm)
+        }
+        knownUsersSearch.accept(searchTerm)
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt:107]: If i search for a query, then change during loading, will it cancel the current query to process the next one?

**Review by BillCarsonFr — CHANGES_REQUESTED** [review]
Quick review, i need to test it

File: vector/src/main/java/im/vector/app/features/invite/InviteUsersToRoomActivity.kt (hunk 1 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/invite/InviteUsersToRoomActivity.kt]: Ups, fixed. Permission logic moved to UserListFragment.

File: vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt (hunk 2 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt]: Done, thanks.

File: vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt (hunk 3 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt:105]: Done.

**Comment by bmarty** [comment]
We will iterate on those screens later on Element Android

**Pull request closed without merging**